File: 04_Haskell/workers/src/Configuration.hs
module Configuration
  ( Configuration (..),
    fromArgs,
  )
where

import Text.Read (readMaybe)

data Configuration = Configuration
  { messageCount :: Int,
    threadCount :: Int
  }
  deriving (Show, Eq)

fromArgs :: [String] -> Either String Configuration
fromArgs args
  | length args < 4 = Left "Invalid input. Pass required arguments: '--message-count', '--thread-count'."
  | otherwise = parse args (Configuration 0 0)

parse :: [String] -> Configuration -> Either String Configuration
parse [] cfg
  | messageCount cfg > 0 && threadCount cfg > 0 = Right cfg
  | otherwise = Left "Invalid input. Pass required arguments: '--message-count', '--thread-count'."
parse ("--message-count" : n : rest) cfg =
  case readMaybe n of
    Nothing -> Left $ "Invalid value for --message-count: " ++ n
    Just v -> parse rest cfg {messageCount = v}
parse ("--thread-count" : n : rest) cfg =
  case readMaybe n of
    Nothing -> Left $ "Invalid value for --thread-count: " ++ n
    Just v -> parse rest cfg {threadCount = v}
parse (_ : rest) cfg = parse rest cfg

File: 06_Kotlin/workers/src/main/kotlin/Configuration.kt
package org.andreastme.langpermonth

private const val MessageCountArg = "--message-count"
private const val ThreadCountArg = "--thread-count"

sealed interface Configuration {
    companion object {
        fun create(args: Array<String>): Configuration {
            if (args.size < 4) {
                println("Invalid input. Pass required arguments: '$MessageCountArg', '$ThreadCountArg'.")
                return Invalid
            }

            var messageCount = 0
            var threadCount = 0

            for (i in args.indices step 2) {
                when (args[i]) {
                    MessageCountArg -> messageCount = args[i + 1].toIntOrNull() ?: 0
                    ThreadCountArg -> threadCount = args[i + 1].toIntOrNull() ?: 0
                }
            }

            return if (messageCount > 0 && threadCount > 0) Valid(messageCount, threadCount) else Invalid
        }
    }

    data class Valid(val messageCount: Int, val threadCount: Int) : Configuration
    data object Invalid : Configuration
}